partition_snapshot_reader.hh

```diff
@@ -89,14 +114,15 @@ class partition_snapshot_flat_reader : public flat_mutation_reader::impl, public
             _clustering_rows.clear();
             _range_tombstones.clear();
 
-            rows_entry::tri_compare rows_cmp(_schema);
+            rows_entry::tri_compare rows_cmp(*_snapshot_schema);
             for (auto&& v : _snapshot->versions()) {
-                mutation_partition::rows_type::const_iterator cr_end = v.partition().upper_bound(_schema, ck_range);
-                auto cr = [&] () -> mutation_partition::rows_type::const_iterator {
+                rows_iter_type cr_end = make_iterator(v.partition().upper_bound(*_snapshot_schema, ck_range));
+
+                auto cr = [&] () -> rows_iter_type {
                     if (last_row) {
-                        return v.partition().clustered_rows().upper_bound(*last_row, rows_cmp);
+                        return make_iterator(v.partition().clustered_rows().upper_bound(*last_row, rows_cmp));
                     } else {
-                        return v.partition().lower_bound(_schema, ck_range);
+                        return make_iterator(v.partition().lower_bound(*_snapshot_schema, ck_range));
```

Comment:
I don't understand this sudden switch to using the snapshot schema for comparing positions with the queried range. What is the motivation here?

In the commit message you explain:
> The reversing partition snapshot reader maintains two schemas - one that
is the reversed schema for the output, and the other one (forward one),
which is used to access the memtable tree (**which needs to be the same as
the schema used to create memtable**)

but that cannot possibly be true - previously (forward) queries used the query schema here for comparisons inside the memtable and it worked.

And I don't think the new approach is correct. Suppose that the snapshot was made using an old schema, then there was a schema upgrade, we're querying the memtable now, so `ck_range` is using the new schema while entries in the snapshot use the old schema. If I understand correctly, the new schema understands old schema entries - so it is able to compare the new clustering key range with the entries in the snapshot. But the snapshot schema does not understand clustering key ranges in the new format.


@denesb could you confirm / deny / correct me?

Reply:
I looked further down and it turns out that the schema passed to this reader was actually the snapshot schema all along (i.e. `_schema` is also the snapshot schema, not the query schema - but for reversed queries it is reversed...). The API of these snapshot reader factory functions is super weird - why is one of them not getting the schema parameter (which is good - because it also takes the snapshot so it can obtain the schema on its own) while the other does (which is bad - because it also takes the snapshot so it can obtain the schema on its own, and it's always being passed that snapshot schema!)

Anyway, summarizing: for reversed queries, `_schema` is the reversed snapshot schema, and `_snapshot_schema` is the non-reversed snapshot schema.

Reply:
I'm not sure what you're talking about:
```cpp
    if (is_reversed) {
        schema_ptr rev_snp_schema = snp->schema()->make_reversed();
        return make_partition_snapshot_flat_reader<true, partition_snapshot_read_accounter>(std::move(rev_snp_schema), std::move(permit), std::move(dk), std::move(crr), std::move(snp), digest_requested, region, read_section, pointer_to_container, fwd, memtable);
    } else {
        return make_partition_snapshot_flat_reader<false, partition_snapshot_read_accounter>(snp->schema(), std::move(permit), std::move(dk), std::move(crr), std::move(snp), digest_requested, region, read_section, pointer_to_container, fwd, memtable);
    }
```
Here, `make_partition_snapshot_flat_reader` doesn't always take the same `schema` as the first argument. 

At the same time, I agree that `_schema` and `_snapshot_schema` could be differentiated better in wording.

Reply:
@kbr- schema upgrades are not relevant for keys (partition or clustering), these can not be altered.